src/main/java/com/gooddata/warehouse/WarehouseService.java

```diff
@@ -513,35 +515,72 @@ public WarehouseS3Credentials getWarehouseS3Credentials(final Warehouse warehous
             throw new GoodDataException("Empty response when S3 credentials POSTed to API");
         }
 
-        return new PollResult<>(this,
-                new AbstractPollHandler<WarehouseTask, WarehouseS3Credentials>
-                        (task.getPollUri(), WarehouseTask.class, WarehouseS3Credentials.class) {
-
-                    @Override
-                    public boolean isFinished(ClientHttpResponse response) throws IOException {
-                        return HttpStatus.CREATED.equals(response.getStatusCode());
-                    }
-
-                    @Override
-                    public void handlePollResult(WarehouseTask pollResult) {
-                        try {
-                            final WarehouseS3Credentials created = restTemplate
-                                    .getForObject(pollResult.getWarehouseS3CredentialsUri(),
-                                            WarehouseS3Credentials.class);
-                            setResult(created);
-                        } catch (GoodDataException | RestClientException e) {
-                            throw new GoodDataException(
-                                    "S3 credentials added to warehouse, but can't get them back, uri: "
-                                            + pollResult.getWarehouseUserUri(), e);
-                        }
-                    }
-
-                    @Override
-                    public void handlePollException(final GoodDataRestException e) {
-                        throw new GoodDataException("Unable to add S3 credentials to warehouse, uri: "
-                                + warehouse.getUri(), e);
-                    }
-                });
+        return new PollResult<>(this, getS3PollHandler(warehouse, s3Credentials, task, "add"));
+    }
+
+
+    /**
+     * update S3 credentials in the Warehouse
+     *
+     * @param warehouse     warehouse to add credentials to
+     * @param s3Credentials the credentials to update
+     * @return updated credentials (not null)
+     */
+    public FutureResult<WarehouseS3Credentials> updateS3CredentialsToWarehouse(final Warehouse warehouse,
```

Comment:
unfortunately it is still needed, because `WarehouseS3Credentials` do not contain any link to `Warehouse`